### ISSUE TITLE
Support modules

### ISSUE THREAD

**Issue by joente** [thingsdb/ThingsDB#141]
Allow ThingsDB to be extended with modules. To add/configure the modules, some new functions are required.

Note that this requires isseu #140.

# Configuration/Environment variable

* `THINGSDB_MODULES_PATH`: Path where to find modules on this node. Defaults to `/usr/lib/thingsdb/modules`

# Future call

To use a module, the future must be used with a `thing` as first argument. This thing must have a `module` property and may have a `deep` property to tell the module how "deep" the argument should be packed for the module. The default `deep` value is `1` if not explicitly given.

For example:
```
// deep 2, otherwise "data" is not being packed
future({
    module: 'SAMPLE',
    deep: 2,
    data: {
        key: 'value'
    }
});
```

## has_module

Returns `true` if the module exists, this does *not* mean that the module can be used since it might be restricted to another scope or the module may not be loaded at all on this, or any node. The function should return `true` anyway so it can be used in the ThingsDB scope to at least check if the module exists.

Scope: `any`

```
has_module(name);
```

Argument | Type | Description 
------------- | ------ | ---------------
`name` | str (required) | Name of the module.

Returns: True/False

## new_module

Scope: `/thingsdb`

```
new_module(name, file, [conf]);
```

Argument | Type | Description 
------------- | ------ | ---------------
`name` | str (required) | Name of the module.
`file` | str (required) | File name for the module. Use `list_modules` in the node scope for a list of available modules.
`conf` | thing/nil (optional) | Configuration for the module. If no configuration is required, this argument must explicitly be set to `nil`. The configuration will be packed using a `deep` value of `2`. This value is fixed and should be sufficient to configure the module.

Return value: `nil`

## module_info

Scope: `/thingsdb`, `/collection`, `/node`

```
module_info(name)
```

Argument | Type | Description 
------------- | ------ | ---------------
`name` | str (required) | Name of the module.

Return value: Info about the module. The return value is a bit different depending on the scope

Value | Description
------- | ---------------
`name` | Name of the module
`file` | File used by this module
`status` | Status of the module. This might be `running` or `error: ....`
`conf` | Configuration for the module. Only visible  with `EVENT` privileges.
`scope` | A module might be restricted to a single scope. When `nil`, this module is available in all scopes.
`tasks` | Number of running tasks. Only visible in a `@node` scope.
`restarts` | Number of unplanned module restarts.

Example return value:

```
{
    "name": "MYSQL",
    "file": "/usr/lib/thingsb/modules/mysql",
    "status": "running",
    "created_at": "...",
    "conf": {
        "userame": "foo",
        "password": "bar",
        "database": "xxx",
        "host": "somehost_or_ip",
        "port": 1234
    },
    "scope": "@collection:stuff",
    "tasks": 0
}
```

**Comment by joente** [comment]
## Function TODO list:

- [ ] cancelled_err
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] else
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] future
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] has_module
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] is_future
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] module_info
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] modules_info
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] new_module
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] rename_module
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] restart_module
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] set_module_conf
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] set_module_scope
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI
- [ ] then
  - [x] Test code
  - [x] Documentation
  - [x] Code
  - [ ] Syntax Highlighting
    - [ ] VSCode
    - [ ] Highlight.js
    - [ ] Create issue for ThingsGUI

**Comment by joente** [comment]
Implemented in **v0.10.2-alpha-6**
